### PR TITLE
feat: send hotkey_detect_request in ddl_client

### PR DESCRIPTION
**Pull request by Smityz**
In the previous design, `hotkey_detect_request` is set in the rrdb_thrift. In this PR, I will refactor it in the replication.thrift

File change:
pegasus: [rrdb.thrift](https://github.com/apache/incubator-pegasus/blob/master/src/idl/rrdb.thrift#L294) -> replication.thrift

### CODE REVIEW

File: include/dsn/dist/replication/replication_ddl_client.h [hunk 1]
```diff
@@ -274,10 +278,31 @@ class replication_ddl_client
...
+            if (err == dsn::ERR_OK) {
+                resp = rpc.response();
+            } else {
+                err = code;
+            }
+        });
+        tracker.wait_outstanding_tasks();
+        return err;
+    }
+
```
Comment (Shuo-Jia) [inline include/dsn/dist/replication/replication_ddl_client.h:300]: So you command would support send to one node meanwhile? can you use `call_rpcs_async` directly which can send one or more node meanwhile

File: include/dsn/dist/replication/replication_ddl_client.h [hunk 2]
```diff
@@ -274,10 +278,31 @@ class replication_ddl_client
         }
     }
 
+    /// Send request to single replica server synchronously.
+    template <typename TRpcHolder, typename TResponse = typename TRpcHolder::response_type>
+    error_code call_rpc_async(dsn::rpc_address address,
```
Comment (Shuo-Jia) [inline include/dsn/dist/replication/replication_ddl_client.h:283]: It's my mistake: https://github.com/XiaoMi/rdsn/blob/1755c53e455443815b4fe5940f95a1f0994b3f1c/include/dsn/dist/replication/replication_ddl_client.h#L247-L250

They should be named `XXXX_sync`, you can help me modify it

File: src/replication.thrift [hunk 3]
```diff
@@ -1071,6 +1071,33 @@ struct query_bulk_load_response
...
+    2: hotkey_detect_action action
+    3: dsn.gpid pid;
+}
+
+struct hotkey_detect_response {
+    // Possible error:
+    // - ERR_OK: start/stop hotkey detect succeed
+    // - ERR_SERVICE_NOT_FOUND: wrong rpc type
+    // - ERR_SERVICE_ALREADY_EXIST: hotkey detection is running now
+    1: i32 err;
```
Comment (hycdong) [inline src/replication.thrift:1097]: The type of `err` should be `dsn.error_code`

File: include/dsn/dist/replication/replication_ddl_client.h (hunk 1 above)
Comment (Smityz) [reply include/dsn/dist/replication/replication_ddl_client.h:300]: I have tried to code like this, but it is very complicated.
I think it's ok to add an interface to support this function.

File: include/dsn/dist/replication/replication_ddl_client.h [hunk 4]
```diff
@@ -267,17 +271,38 @@ class replication_ddl_client
...
+            call_rpcs_sync(rpcs, retry_resps, reply_thread_hash, false);
             for (auto &resp : retry_resps) {
                 resps.emplace(resp.first, std::move(resp.second));
             }
         }
     }
 
+    /// Send request to single replica server synchronously.
+    template <typename TRpcHolder, typename TResponse = typename TRpcHolder::response_type>
+    error_code call_rpc_sync(dsn::rpc_address address,
```
Comment (levy5307) [inline include/dsn/dist/replication/replication_ddl_client.h:283]: const dsn::rpc_address address

File: src/client/replication_ddl_client.cpp [hunk 5]
```diff
@@ -1580,5 +1580,15 @@ replication_ddl_client::query_bulk_load(const std::string &app_name)
     return call_rpc_sync(query_bulk_load_rpc(std::move(req), RPC_CM_QUERY_BULK_LOAD_STATUS));
 }
 
+error_code replication_ddl_client::detect_hotkey(const dsn::rpc_address &target,
+                                                 detect_hotkey_request &req,
+                                                 detect_hotkey_response &resp)
+{
+    auto request = make_unique<detect_hotkey_request>(req);
+    detect_hotkey_rpc rpc(std::move(request), RPC_DETECT_HOTKEY);
+    error_code err = call_rpc_sync(target, rpc, resp);
```
Comment (levy5307) [inline src/client/replication_ddl_client.cpp:1589]: `return call_rpc_sync(target, rpc, resp);`

File: src/replica/replica_stub.cpp [hunk 6]
```diff
@@ -2766,6 +2768,19 @@ void replica_stub::on_group_bulk_load(group_bulk_load_rpc rpc)
...
+    const auto &request = rpc.request();
+    auto &response = rpc.response();
+    replica_ptr rep = get_replica(request.pid);
+    if (rep != nullptr) {
+        response.err = ERR_OK;
+    } else {
+        response.err = ERR_SERVICE_NOT_FOUND;
+        response.err_hint = "not find the replica";
+    }
+};
```
Comment (levy5307) [inline src/replica/replica_stub.cpp:2783]: why did you add a `;` behind func?

File: src/replica/replica_stub.h [hunk 7]
```diff
@@ -206,6 +207,9 @@ class replica_stub : public serverlet<replica_stub>, public ref_counter
     // This function is used for partition split error handler
     void split_replica_error_handler(gpid pid, local_execution handler);
 
+    // Hotkey detection
```
Comment (levy5307) [inline src/replica/replica_stub.h:210]: I think the comment doesn't make sense, so you can remove it.

File: include/dsn/dist/replication/replication_ddl_client.h [hunk 8]
```diff
@@ -267,17 +271,38 @@ class replication_ddl_client
...
+                             TRpcHolder rpc,
+                             /*out*/ TResponse &resp)
+    {
+        dsn::task_tracker tracker;
+        error_code err = ERR_UNKNOWN;
+        rpc.call(address, &tracker, [&err, &resp, rpc](error_code code) mutable {
+            err = code;
+            if (err == dsn::ERR_OK) {
+                resp = rpc.response();
+            } else {
```
Comment (levy5307) [inline include/dsn/dist/replication/replication_ddl_client.h:293]: remove the `else` branch

**Comment by neverchanje** [comment]
Remove include/dsn/git_commit.h. This file was deleted previously.

File: src/runtime/security/security_types.h [hunk 9]
```diff
@@ -147,8 +147,7 @@ inline std::ostream &operator<<(std::ostream &out, const negotiation_response &o
     obj.printTo(out);
     return out;
 }
-
-} // namespace security
-} // namespace dsn
+}
+} // namespace
```
Comment (neverchanje) [inline src/runtime/security/security_types.h:151]: security.thrift is not modified. Strip out these files in your PR.

File: src/replica/replica_stub.cpp [hunk 10]
```diff
@@ -2766,6 +2768,19 @@ void replica_stub::on_group_bulk_load(group_bulk_load_rpc rpc)
...
+// TODO: (Tangyanzhao) implement it later
+void replica_stub::on_detect_hotkey(detect_hotkey_rpc rpc)
+{
+    const auto &request = rpc.request();
+    auto &response = rpc.response();
+    replica_ptr rep = get_replica(request.pid);
+    if (rep != nullptr) {
+        response.err = ERR_OK;
+    } else {
+        response.err = ERR_SERVICE_NOT_FOUND;
```
Comment (hycdong) [inline src/replica/replica_stub.cpp:2780]: We usually use `ERR_OBJECT_NOT_FOUND` indicating can not found repica in this stub. Besides, add `derror_f` for this error, add replica pid infomation in err_hint.

File: src/replication.thrift [hunk 11]
```diff
@@ -1071,6 +1071,33 @@ struct query_bulk_load_response
...
+struct detect_hotkey_request {
+    1: hotkey_type type
+    2: detect_action action
+    3: dsn.gpid pid;
+}
+
+struct detect_hotkey_response {
+    // Possible error:
+    // - ERR_OK: start/stop hotkey detect succeed
+    // - ERR_SERVICE_NOT_FOUND: wrong rpc type
```
Comment (hycdong) [inline src/replication.thrift:1095]: Use `ERR_OBJECT_NOT_FOUND` to show replica not found

File: include/dsn/dist/replication/replication_ddl_client.h (hunk 1 above)
Comment (hycdong) [reply include/dsn/dist/replication/replication_ddl_client.h:294]: I agree with @Shuo-Jia comment, you can use function `call_rpcs_sync`.

File: src/client/replication_ddl_client.cpp [hunk 12]
```diff
@@ -1580,5 +1580,19 @@ replication_ddl_client::query_bulk_load(const std::string &app_name)
     return call_rpc_sync(query_bulk_load_rpc(std::move(req), RPC_CM_QUERY_BULK_LOAD_STATUS));
 }
 
+error_code replication_ddl_client::detect_hotkey(const dsn::rpc_address &target,
```
Comment (Shuo-Jia) [inline src/client/replication_ddl_client.cpp:1583]: Does the commend support send multi nodes？if support, you should pass  `targets`

File: src/replica/replica_stub.cpp [hunk 13]
```diff
@@ -2766,6 +2769,25 @@ void replica_stub::on_group_bulk_load(group_bulk_load_rpc rpc)
         response.err = ERR_OBJECT_NOT_FOUND;
     }
 }
```
Comment (levy5307) [inline src/replica/replica_stub.cpp:2771]: add a blank line

File: src/client/replication_ddl_client.cpp (hunk 12 above)
Comment (Smityz) [reply src/client/replication_ddl_client.cpp:1583]: I think there is no need to support sending to multi nodes right now.

**Review by hycdong — APPROVED** [review]

File: src/client/replication_ddl_client.cpp [hunk 14]
```diff
@@ -1580,5 +1580,19 @@ replication_ddl_client::query_bulk_load(const std::string &app_name)
...
+                                                 detect_hotkey_request &req,
+                                                 detect_hotkey_response &resp)
+{
+    std::map<dsn::rpc_address, detect_hotkey_rpc> detect_hotkey_rpcs;
+    auto request = make_unique<detect_hotkey_request>(req);
+    detect_hotkey_rpcs.emplace(target, detect_hotkey_rpc(std::move(request), RPC_DETECT_HOTKEY));
+    std::map<dsn::rpc_address, error_with<detect_hotkey_response>> resps;
+    call_rpcs_sync(detect_hotkey_rpcs, resps);
+    auto err_info = resps.begin()->second.get_error().code();
```
Comment (levy5307) [inline src/client/replication_ddl_client.cpp:1592]:
```
resp = resps.begin()->second.get_value();
return resps.begin()->second.get_error().code();
```

**Review by levy5307 — APPROVED** [review]

**Review by hycdong — APPROVED** [review]
